<a href="https://colab.research.google.com/github/lotusist/MachineLearning-DeepLearning/blob/main/hw1_multilinearreg_logit.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# MLDL1 2022 Spring
## Assignment 1

# 0. Setting Datasets

In [ ]:
import numpy as np
import pandas as pd
from sklearn.linear_model import LinearRegression, LogisticRegression
import statsmodels.api as sm

In [ ]:
_path = 'C:/Users/gsds/Desktop/MLDL1'
data_list = ['Carseats', 'Default']

In [ ]:
cs = pd.read_csv(_path + '/' + data_list[0] + '.csv', low_memory = False)
df = pd.read_csv(_path + '/' + data_list[1] + '.csv', low_memory = False)

# 1. Carseats

## (a) Fit a multiple linear regression model to predict Sales using Price, Urban,and US. Report the $ R^2 $ of the model.
 - The $ R^2 $ of the model is **0.239**.

In [ ]:
# Slice the columns to analyze from the original dataset
dt = cs[['Sales', 'Price', 'Urban', 'US']]
dt.info()
dt.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 400 entries, 0 to 399
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Sales   400 non-null    float64
 1   Price   400 non-null    int64  
 2   Urban   400 non-null    object 
 3   US      400 non-null    object 
dtypes: float64(1), int64(1), object(2)
memory usage: 12.6+ KB


Sales  Price Urban   US
0   9.50    120   Yes  Yes
1  11.22     83   Yes  Yes
2  10.06     80   Yes  Yes
3   7.40     97   Yes  Yes
4   4.15    128   Yes   No

In [ ]:
# create dummy variables for categorical values
_cate = ['Urban', 'US']
for i in _cate:
  dt[i] = dt[i].map({'Yes' : 1,
                     'No': 0})
dt.info()
dt.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 400 entries, 0 to 399
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Sales   400 non-null    float64
 1   Price   400 non-null    int64  
 2   Urban   400 non-null    int64  
 3   US      400 non-null    int64  
dtypes: float64(1), int64(3)
memory usage: 12.6 KB


C:\Users\gsds_lsj\AppData\Local\Temp\ipykernel_12836\3370123073.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dt[i] = dt[i].map({'Yes' : 1,
C:\Users\gsds_lsj\AppData\Local\Temp\ipykernel_12836\3370123073.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dt[i] = dt[i].map({'Yes' : 1,


Sales  Price  Urban  US
0   9.50    120      1   1
1  11.22     83      1   1
2  10.06     80      1   1
3   7.40     97      1   1
4   4.15    128      1   0

In [ ]:
# check the correlation between the qualitative variables
check_corr = dt[['Urban', 'US']]
print(check_corr.corr())

          Urban        US
Urban  1.000000  0.047085
US     0.047085  1.000000


In [ ]:
# Split predictors and response variable
X = dt[['Price', 'Urban', 'US']]
y = dt['Sales']

In [ ]:
# Fit the multiple linear regression model
Xa = sm.add_constant(X)
lr1= sm.OLS(y, Xa)
fit_1= lr1.fit()

In [ ]:
# Overview the result of the modeling
print(fit_1.summary())

                            OLS Regression Results                            
Dep. Variable:                  Sales   R-squared:                       0.239
Model:                            OLS   Adj. R-squared:                  0.234
Method:                 Least Squares   F-statistic:                     41.52
Date:                Wed, 06 Apr 2022   Prob (F-statistic):           2.39e-23
Time:                        21:10:40   Log-Likelihood:                -927.66
No. Observations:                 400   AIC:                             1863.
Df Residuals:                     396   BIC:                             1879.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         13.0435      0.651     20.036      0.0

## (b) Provide an interpretation of each coefficient in the model.

 - The coefficients of the model are shown in (a) OLS Regression Results above. 
  - $ \hat{\beta_{0}} $ : 13.0435
  - $ \hat{\beta_{Price}} $ : -0.0545
  - $ \hat{\beta_{Urban}} $ : -0.0219
  - $ \hat{\beta_{US}} $ : 1.2006
 - The coefficient of the quantitative predictor 'Price' is -0.0545, which quantifies the average effect of one unit increase in Price on Sales while other predictors are fixed. That is, **one unit increase of Price is associated with the drop of 0.0545 unit in Sales.** 
 - The coefficient of the qualitative predictor 'Urban' is -0.0219, which means **the Sales in urban area may be lower than those not in urban area by -0.0219 unit on average**, if other predictors are fixed.
 - The coefficient of the qualitative predictor 'US' is 1.2006, which means **the Sales in the US is 1.2006 unit higher than those not in the US**, if other predictors are fixed.
 - The intecept is the baseline, which is **the average of the Sales where the Price is zero, and neither in the US nor in the Urban.**

## (c) Write out the model in equation form, being careful to handle the qualitative variables properly.
$$ Y_{Sales} = 13.0435 -0.0545X_{Price} -0.0219X_{Urban} +1.2006X_{US} $$ 


## (d) For which predictor variable j can you reject the null hypothesis $ H_{0} : β_{j} =0 $ ?

 - The P-value ($P>\vert t\vert $) shown in (a) OLS Regresion Results above indicates the probability to accept the null hypothesis.
 - **$ X_{Price} $ and $ X_{US} $** reject the null hypothesis, since the P-value is low enough (close to 0).
 - $ X_{Urban} $, however, whose P-value is high(0.936), does not reject the null hypothesis. 

## (e) On the basis of your response to the previous question, fit a smaller model that only uses the predictors for which there is evidence of association with the outcome.

 - The new model contains only 2 predictors $ X_{Price} $ and $ X_{US} $, whose P-values are close to 0. That is, these predictors are significant.
 - The results of the new smaller model is below.

In [ ]:
# Fit the smaller model and overview the result of the modeling

X = dt[['Price', 'US']]
y = dt['Sales']
Xe = sm.add_constant(X)
lr2 = sm.OLS(y, Xe)
fit_2= lr2.fit()
print(fit_2.summary())

                            OLS Regression Results                            
Dep. Variable:                  Sales   R-squared:                       0.239
Model:                            OLS   Adj. R-squared:                  0.235
Method:                 Least Squares   F-statistic:                     62.43
Date:                Wed, 06 Apr 2022   Prob (F-statistic):           2.66e-24
Time:                        21:10:57   Log-Likelihood:                -927.66
No. Observations:                 400   AIC:                             1861.
Df Residuals:                     397   BIC:                             1873.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         13.0308      0.631     20.652      0.0

## (f) How well do the models in Part (a) and Part (e) fit the data? Compare the two models. Can you say one model is better than the other based on $ R^{2} $ ?

 - The $R^{2}$ of the first regression model (a) is 0.239, and that of the smaller model (e) is same, **which does not give any implications on the improvement of the model performance.**
 - However, F-statistic, which also indicates the overall performance of the model, is higher in the model (e). Thus, **the model (e) is better than model (a).** 
 - In general, **small models have more advantages than complex models in terms of the interpretability and predictive accurracy** on the datapoints unobserved yet. 

## (g) Using the model from (e), obtain 95% confidence intervals for the coefficient(s).

 - the 95% confidence interval for each coefficient can be calculated by the formula below.
  - $ \hat{\beta} \pm 2*SE(\hat{\beta}) $
 - The confidence interval of the coefficient of $ X_{Price} $ (or, $ \hat{\beta_{Price}} $)is **between -0.065 and -0.044.**
 - The confidence interval of the coefficient of $ X_{US} $ (or, $ \hat{\beta_{US}} $)is **between 0.684 and 1.716.**
 - The confidence interval of intercept(or, $ \hat{\beta_{0}} $) is **between 11.769 and 14.293.**

In [ ]:
def conf_interval(_coeff: float, _se:float, _lambda:float) -> None:
  start = _coeff - _lambda*_se
  end = _coeff + _lambda*_se
  return print([round(start,3), round(end,3)])

In [ ]:
interc_coeff = 13.0308
interc_se = 0.631

price_coeff =  -0.0545
price_se = 0.005

us_coeff = 1.1996
us_se = 0.258

conf_interval(price_coeff, price_se, 2)
conf_interval(us_coeff, us_se, 2)
conf_interval(interc_coeff, interc_se, 2)

[-0.065, -0.044]
[0.684, 1.716]
[11.769, 14.293]


# 2. Default

## (a) Fit a logistic regression model that uses income and balance to predict default using the validation set approach, and estimate the test error of this model. In order to do this, you must perform the following steps:
 - Split the sample set into a training set and a validation set.
 - Fit a multiple logistic regression model using only the training observations.
 - Obtain a prediction of default status for each individual in the validation set by computing the posterior probability of default for that individual, and classifying the individual to the default category if the posterior probability is greater than 0.5.
 - Compute the validation set error, which is the fraction of the observations in the validation set that are misclassified.

In [ ]:
# Set random seed 
np.random.seed(25212)

In [ ]:
# Create dummy variables for categorical variables
_cate2 = ['default', 'student']
for i in _cate2:
  df[i] = df[i].map({'Yes' : 1,
                     'No': 0})
df.info()
df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 4 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   default  10000 non-null  int64  
 1   student  10000 non-null  int64  
 2   balance  10000 non-null  float64
 3   income   10000 non-null  float64
dtypes: float64(2), int64(2)
memory usage: 312.6 KB


default  student      balance        income
0        0        0   729.526495  44361.625074
1        0        1   817.180407  12106.134700
2        0        0  1073.549164  31767.138947
3        0        0   529.250605  35704.493935
4        0        0   785.655883  38463.495879

In [ ]:
# Slice the columns to analyze from the original dataset
dff = df[['default', 'income', 'balance']]

In [ ]:
# check the correlation between the qualitative predictors
check_corr2 = dff[['income', 'balance']]
print(check_corr2.corr())

           income   balance
income   1.000000 -0.152243
balance -0.152243  1.000000


 - Split the sample set randomly shuffled into a training set and a validation set.

In [ ]:
train, validate = np.split(dff.sample(frac=1, random_state=25212), [int(0.5*len(dff))])

In [ ]:
# Split predictors and response variable 
X_train = train[['income','balance']]
y_train = train['default']
X_validate = validate[['income','balance']]
y_validate = validate['default']

In [ ]:
print(X_train.shape)
print(y_train.shape)
print(X_validate.shape)
print(y_validate.shape)

(5000, 2)
(5000,)
(5000, 2)
(5000,)


 - Fit a multiple logistic regression model using only the training observations.

In [ ]:
X_train = sm.add_constant(X_train)
X_validate = sm.add_constant(X_validate)
lg = sm.Logit(y_train, X_train)
fit_lg= lg.fit()
print(fit_lg.summary())

Optimization terminated successfully.
         Current function value: 0.076068
         Iterations 10
                           Logit Regression Results                           
Dep. Variable:                default   No. Observations:                 5000
Model:                          Logit   Df Residuals:                     4997
Method:                           MLE   Df Model:                            2
Date:                Wed, 06 Apr 2022   Pseudo R-squ.:                  0.4680
Time:                        21:12:44   Log-Likelihood:                -380.34
converged:                       True   LL-Null:                       -714.94
Covariance Type:            nonrobust   LLR p-value:                4.836e-146
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const        -11.8739      0.640    -18.540      0.000     -13.129     -10.619
income      2.383e-05   7.09

 - Obtain a prediction of default status for each individual in the validation set by computing the posterior probability of default for that individual. 

In [ ]:
y_hat = fit_lg.predict(X_validate)

In [ ]:
pred_actual= pd.DataFrame(data = y_hat, columns=['prob'])
pred_actual.insert(loc = 1, column = 'default_predicted', value = None)

 - Classify the individual to the default category if the posterior probability is greater than 0.5.

In [ ]:
def convert_to_default(x: float) -> int:
  if x > 0.5:
    return 1
  else:
    return 0

In [ ]:
for i in pred_actual.index:
  pred_actual['default_predicted'][i] = convert_to_default(pred_actual['prob'][i])

C:\Users\gsds_lsj\AppData\Local\Temp\ipykernel_12836\2368642204.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  pred_actual['default_predicted'][i] = convert_to_default(pred_actual['prob'][i])


In [ ]:
# The probability of default and the predicted default category of each individual in the validation set
pred_actual

prob default_predicted
6210  0.001774                 0
5693  0.024642                 0
3425  0.000011                 0
1455  0.001609                 0
5327  0.000441                 0
...        ...               ...
1211  0.001243                 0
5818  0.032595                 0
2176  0.000741                 0
2281  0.002434                 0
9595  0.100588                 0

[5000 rows x 2 columns]

In [ ]:
# the number of the individuals who are predicted to be default is 66
pred_actual['default_predicted'].value_counts()

0    4934
1      66
Name: default_predicted, dtype: int64

 - Compute the validation set error, which is the fraction of the observations in the validation set that are misclassified.
  - **Misclassification error : 0.027**
  - **False positive rate : 0.003**
  - **False_negative_rate : 0.702** 
  - False negative rate is too high; that is to say, **this model fails to predict individuals to be default actually.** This issue may result from the imbalanced ratio of the labels. 

 - Performance Metrics
  - Recall(or True positive rate, sensitivity) : 0.298
  - Precision : 0.773
  - AUC : 0.947
  - The recall is too low, which indicates the same issue as above.

In [ ]:
# The predicted default category and the actual default status of each individual in the validation set
pred_actual.insert(loc =2, column = 'default_actual', value = y_validate)
pred_actual

prob default_predicted  default_actual
6210  0.001774                 0               0
5693  0.024642                 0               0
3425  0.000011                 0               0
1455  0.001609                 0               0
5327  0.000441                 0               0
...        ...               ...             ...
1211  0.001243                 0               0
5818  0.032595                 0               0
2176  0.000741                 0               0
2281  0.002434                 0               0
9595  0.100588                 0               0

[5000 rows x 3 columns]

In [ ]:
# the number of the individuals who are actually default is 171
pred_actual['default_actual'].value_counts()

0    4829
1     171
Name: default_actual, dtype: int64

In [ ]:
# Construct a confusion matrix to calculate the validation set error
confusion_matrix = [[0,0],[0,0]]
for i in pred_actual.index:
  if pred_actual['default_predicted'][i] == 0:
    if pred_actual['default_actual'][i] == 0:
      confusion_matrix[0][0] += 1 # True Negative
    else :
      confusion_matrix[1][0] += 1 # False Negative
  else :
    if pred_actual['default_actual'][i] == 0:
      confusion_matrix[0][1] += 1 # False Positive
    else :
      confusion_matrix[1][1] += 1 # True Positive
confusion_matrix

[[4814, 15], [120, 51]]

In [ ]:
# Calculate the misclassification error of prediction model
a_misclassification_error = (confusion_matrix[0][1] + confusion_matrix[1][0]) / (confusion_matrix[0][0] + confusion_matrix[0][1] + confusion_matrix[1][0] + confusion_matrix[1][1])
print(a_misclassification_error)

0.027


In [ ]:
# Other performance metrics to consider the imbalanced ratio of the labels('default')
a_false_positive_rate = confusion_matrix[0][1] / (confusion_matrix[0][0] + confusion_matrix[0][1]) 
a_false_negative_rate = confusion_matrix[1][0] / (confusion_matrix[1][0] + confusion_matrix[1][1])
a_true_positive_rate = confusion_matrix[1][1] / (confusion_matrix[1][0]+confusion_matrix[1][1])# sensitivity, or recall
a_precision = confusion_matrix[1][1] / (confusion_matrix[0][1]+confusion_matrix[1][1])
a_specificity = 1-a_false_positive_rate

print("false positive rate : ", round(a_false_positive_rate,3))
print("false negative rate : ", round(a_false_negative_rate,3))
print("True positive rate (Recall, Power, Sensitivity): ", round(a_true_positive_rate,3))
print("precision : ", round(a_precision,3))
print("specificity : ", round(a_specificity,3))

false positive rate :  0.003
false negative rate :  0.702
True positive rate (Recall, Power, Sensitivity):  0.298
precision :  0.773
specificity :  0.997


In [ ]:
# AUC
from sklearn import metrics
y_actual = np.array(pred_actual['default_actual'])
fpr, tpr, thresholds = metrics.roc_curve(y_actual, y_hat)
round(metrics.auc(fpr, tpr),3)

0.947

## (b) Provide an interpretation of each coefficient in the trained model.


 - The coefficients of the model are shown in (a) the Logit Regression Result above.
  - $ \hat{\beta_{0}} $ : -11.8739
  - $ \hat{\beta_{income}} $ : 2.383e-05
  - $ \hat{\beta_{balance}} $ : 0.0058
 - The $ \hat{\beta} $ in logistic regression indicates that how much the log odds of the default changes on average with one unit increase of the predictor, if values of other predictors are fixed.
  - $ \hat{\beta_{income}} $ indicates that **increasing income by one unit changes the log odds slightly by 2.383e-05 on average.** Considering the P-value is low, the income is a significant variable.
  - $ \hat{\beta_{balance}} $ also indicates that **the increasing balance by one unit associated with the increasing change of the log odds by 0.0058 on average.** Since the P-value is close to 0, the balance can be regarded as a significant predictor. 
  - The intercept, or $ \hat{\beta_{0}} $ is **the log odds when both the income and balance are zero.**

## (c) Perform 5-fold cross-validation using the same model as in Part (a). How does the validation error of the cross-validation the differ from to the results in Part (a)?

 - Sample n = len(dataset) / k(=5) datapoints, and split them as the validation set.
 - Other datapoints make up training set.
 - Fit a logistic regression model, calcuate misclassification error metrics each.
 - Repeat the sequence by sampling without replacement for k(=5) times. 
 - Average the errors from k(=5) validation sets and get the validation error of the cross-validation. 

In [ ]:
def c_k_fold_cv(dataset: pd.DataFrame, K: int):
  val_used_rows = []
  valid_rows = []
  
  MisClassError = []
  FalsePosR = []
  FalseNegR = []
  Recall = []
  Precision = []

  N = int(dataset.shape[0]/K)

  for times in range(K):
    _copy = dataset.copy()
    _sample_space = _copy.drop(val_used_rows) # delete the rows which was the validation data previouly in order to avoid overlapping issue on the posterior validation sets
    valid_set = _sample_space.sample(n = N, random_state = 25212) # sample the same size of datapoints for each fold
    for j in valid_set.index:
      val_used_rows.append(j) # register the rows once used as the validation data
      valid_rows.append(j)  # register the rows used as the validation data currently
    train_set = _copy.drop(valid_rows) # delete the rows used as the validation data currently from the train set
    valid_rows.clear() # clear the register memory of the current validation set

    X_train = train_set[['income', 'balance']]
    y_train = train_set['default']
    X_valid = valid_set[['income','balance']]
    y_valid = valid_set['default']
    X_train = sm.add_constant(X_train)
    X_valid = sm.add_constant(X_valid)
    lg_model = sm.Logit(y_train, X_train)
    fit_lg_model = lg_model.fit()
    y_hat = fit_lg_model.predict(X_valid)

    tmp= pd.DataFrame(data = y_hat, columns=['prob'])
    tmp.insert(loc = 1, column = 'default_predicted', value = None)
    for k in tmp.index:
      tmp['default_predicted'][k] = convert_to_default(tmp['prob'][k])
    tmp.insert(loc=2, column = 'default_actual', value = y_valid)

    con_mat = [[0,0],[0,0]]
    for l in tmp.index:
      if tmp['default_predicted'][l] == 0:
        if tmp['default_actual'][l] == 0:
          con_mat[0][0] += 1 
        else :
          con_mat[1][0] += 1 
      else :
        if tmp['default_actual'][l] == 0:
          con_mat[0][1] += 1 
        else :
          con_mat[1][1] += 1 

    misclass_error = (con_mat[0][1] + con_mat[1][0]) / (con_mat[0][0] + con_mat[0][1] + con_mat[1][0] + con_mat[1][1])
    MisClassError.append(misclass_error)
    false_positive_rate = con_mat[0][1] / (con_mat[0][0] + con_mat[0][1])
    FalsePosR.append(false_positive_rate)
    false_negative_rate = con_mat[1][0] / (con_mat[1][0] + con_mat[1][1])
    FalseNegR.append(false_negative_rate)
    recall = con_mat[1][1] / (con_mat[1][0]+con_mat[1][1])
    Recall.append(recall)
    precision = con_mat[1][1] / (con_mat[0][1]+con_mat[1][1])
    Precision.append(precision)

  c_misclass_error = sum(MisClassError) / K
  c_false_positive_rate = sum(FalsePosR) / K
  c_false_negative_rate = sum(FalseNegR) / K
  c_recall = sum(Recall) / K
  c_precision = sum(Precision) / K

  return c_misclass_error, c_false_positive_rate, c_false_negative_rate, c_recall, c_precision

In [ ]:
c_misclass_error, c_false_positive_rate, c_false_negative_rate, c_recall, c_precision = c_k_fold_cv(dff,5)

Optimization terminated successfully.
         Current function value: 0.080578
         Iterations 10


C:\Users\gsds_lsj\AppData\Local\Temp\ipykernel_12836\967960528.py:36: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  tmp['default_predicted'][k] = convert_to_default(tmp['prob'][k])


Optimization terminated successfully.
         Current function value: 0.078300
         Iterations 10


C:\Users\gsds_lsj\AppData\Local\Temp\ipykernel_12836\967960528.py:36: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  tmp['default_predicted'][k] = convert_to_default(tmp['prob'][k])


Optimization terminated successfully.
         Current function value: 0.080147
         Iterations 10


C:\Users\gsds_lsj\AppData\Local\Temp\ipykernel_12836\967960528.py:36: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  tmp['default_predicted'][k] = convert_to_default(tmp['prob'][k])


Optimization terminated successfully.
         Current function value: 0.076528
         Iterations 10


C:\Users\gsds_lsj\AppData\Local\Temp\ipykernel_12836\967960528.py:36: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  tmp['default_predicted'][k] = convert_to_default(tmp['prob'][k])


Optimization terminated successfully.
         Current function value: 0.079108
         Iterations 10


C:\Users\gsds_lsj\AppData\Local\Temp\ipykernel_12836\967960528.py:36: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  tmp['default_predicted'][k] = convert_to_default(tmp['prob'][k])


In [ ]:
print('MISCLASS_ERROR :', round(c_misclass_error,3,), 
      'FPR :', round(c_false_positive_rate,3), 
      'FNR :', round(c_false_negative_rate,3), 
      'RECALL :', round(c_recall,3), 
      'PRECISION :', round(c_precision, 3))

MISCLASS_ERROR : 0.026 FPR : 0.004 FNR : 0.675 RECALL : 0.325 PRECISION : 0.74


 - Check how much the validation error of the 5-fold cross-validation differ from to the results in Part (a).
  - The overall misclassification error has been slightly improved. 
  - **The metrics such as FNR and recall has been improved,** which are associated with predicting the actual default status. 

In [ ]:
comp1 = {'MISCLASS_ERROR': [round(a_misclassification_error,3), round(c_misclass_error,3)], 
        'FPR': [round(a_false_positive_rate,3), round(c_false_positive_rate,3)], 
        'FNR': [round(a_false_negative_rate,3), round(c_false_negative_rate,3)], 
        'RECALL': [round(a_true_positive_rate,3), round(c_recall,3)], 
        'PRCSN': [round(a_precision,3), round(c_precision,3)] }

comparison1 = pd.DataFrame.from_dict(comp1, orient='index')
comparison1.columns = ['model (a)', 'model (c)']
comparison1

model (a)  model (c)
MISCLASS_ERROR      0.027      0.026
FPR                 0.003      0.004
FNR                 0.702      0.675
RECALL              0.298      0.325
PRCSN               0.773      0.740

## (d)  Now consider a logistic regression model that predicts the probability of default using income, balance, and a dummy variable for student. Estimate the test error for this model using the 5-fold cross-validation set approach. Comment on whether or not including a dummy variable for student would lead to a reduction in the test error rate.


In [ ]:
def d_k_fold_cv(dataset: pd.DataFrame, K: int):
  val_used_rows = []
  valid_rows = []
  
  MisClassError = []
  FalsePosR = []
  FalseNegR = []
  Recall = []
  Precision = []

  N = int(dataset.shape[0]/K)

  for times in range(K):
    _copy = dataset.copy()
    _sample_space = _copy.drop(val_used_rows) # delete the rows which was the validation data previouly in order to avoid overlapping issue on the posterior validation sets
    valid_set = _sample_space.sample(n = N, random_state = 25212) # sample the same size of datapoints for each fold
    for j in valid_set.index:
      val_used_rows.append(j) # register the rows once used as the validation data
      valid_rows.append(j)  # register the rows used as the validation data currently
    train_set = _copy.drop(valid_rows) # delete the rows used as the validation data currently from the train set
    valid_rows.clear() # clear the register memory of the current validation set

    X_train = train_set[['income', 'balance', 'student']]
    y_train = train_set['default']
    X_valid = valid_set[['income','balance', 'student']]
    y_valid = valid_set['default']
    X_train = sm.add_constant(X_train)
    X_valid = sm.add_constant(X_valid)
    lg_model = sm.Logit(y_train, X_train)
    fit_lg_model = lg_model.fit()
    y_hat = fit_lg_model.predict(X_valid)

    tmp= pd.DataFrame(data = y_hat, columns=['prob'])
    tmp.insert(loc = 1, column = 'default_predicted', value = None)
    for k in tmp.index:
      tmp['default_predicted'][k] = convert_to_default(tmp['prob'][k])
    tmp.insert(loc=2, column = 'default_actual', value = y_valid)

    con_mat = [[0,0],[0,0]]
    for l in tmp.index:
      if tmp['default_predicted'][l] == 0:
        if tmp['default_actual'][l] == 0:
          con_mat[0][0] += 1 
        else :
          con_mat[1][0] += 1 
      else :
        if tmp['default_actual'][l] == 0:
          con_mat[0][1] += 1 
        else :
          con_mat[1][1] += 1 

    misclass_error = (con_mat[0][1] + con_mat[1][0]) / (con_mat[0][0] + con_mat[0][1] + con_mat[1][0] + con_mat[1][1])
    MisClassError.append(misclass_error)
    false_positive_rate = con_mat[0][1] / (con_mat[0][0] + con_mat[0][1])
    FalsePosR.append(false_positive_rate)
    false_negative_rate = con_mat[1][0] / (con_mat[1][0] + con_mat[1][1])
    FalseNegR.append(false_negative_rate)
    recall = con_mat[1][1] / (con_mat[1][0]+con_mat[1][1])
    Recall.append(recall)
    precision = con_mat[1][1] / (con_mat[0][1]+con_mat[1][1])
    Precision.append(precision)

  c_misclass_error = sum(MisClassError) / K
  c_false_positive_rate = sum(FalsePosR) / K
  c_false_negative_rate = sum(FalseNegR) / K
  c_recall = sum(Recall) / K
  c_precision = sum(Precision) / K

  return c_misclass_error, c_false_positive_rate, c_false_negative_rate, c_recall, c_precision

In [ ]:
d_misclass_error, d_false_positive_rate, d_false_negative_rate, d_recall, d_precision = d_k_fold_cv(df,5)

Optimization terminated successfully.
         Current function value: 0.080056
         Iterations 10


C:\Users\gsds_lsj\AppData\Local\Temp\ipykernel_12836\1622079427.py:36: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  tmp['default_predicted'][k] = convert_to_default(tmp['prob'][k])


Optimization terminated successfully.
         Current function value: 0.077876
         Iterations 10


C:\Users\gsds_lsj\AppData\Local\Temp\ipykernel_12836\1622079427.py:36: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  tmp['default_predicted'][k] = convert_to_default(tmp['prob'][k])


Optimization terminated successfully.
         Current function value: 0.079753
         Iterations 10


C:\Users\gsds_lsj\AppData\Local\Temp\ipykernel_12836\1622079427.py:36: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  tmp['default_predicted'][k] = convert_to_default(tmp['prob'][k])


Optimization terminated successfully.
         Current function value: 0.076220
         Iterations 10


C:\Users\gsds_lsj\AppData\Local\Temp\ipykernel_12836\1622079427.py:36: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  tmp['default_predicted'][k] = convert_to_default(tmp['prob'][k])


Optimization terminated successfully.
         Current function value: 0.078863
         Iterations 10


C:\Users\gsds_lsj\AppData\Local\Temp\ipykernel_12836\1622079427.py:36: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  tmp['default_predicted'][k] = convert_to_default(tmp['prob'][k])


 - Check the effect of dummy variable for 'Student'
  - Including the dummy variable 'Student' **didn't help to reduce the validation error** (,or the estimated test error) model (c), as the overall misclassification error got higher, especially the FNR got worse.
  - In this vein both the recall and precision score got worse in model (d), which including the 'Student' dummy variable.

In [ ]:
comp2 = {'MISCLASS_ERROR': [round(c_misclass_error,3), round(d_misclass_error,3)], 
        'FPR': [round(c_false_positive_rate,3), round(d_false_positive_rate,3)], 
        'FNR': [round(c_false_negative_rate,3), round(d_false_negative_rate,3)], 
        'RECALL': [round(c_recall,3), round(d_recall,3)], 
        'PRCSN': [round(c_precision,3), round(d_precision,3)] }

comparison2 = pd.DataFrame.from_dict(comp2, orient='index')
comparison2.columns = ['model (c)', 'model (d)']
comparison2

model (c)  model (d)
MISCLASS_ERROR      0.026      0.027
FPR                 0.004      0.004
FNR                 0.675      0.680
RECALL              0.325      0.320
PRCSN               0.740      0.731